src/main/java/com/github/mc1arke/sonarqube/plugin/ce/pullrequest/gitlab/GitlabServerPullRequestDecorator.java

```diff
@@ -0,0 +1,379 @@
+                String nextURL = getNextUrl(httpResponse);
+                if (nextURL != null) {
+                    LOGGER.info("Getting next page");
+                    discussions.addAll(getCommitDiscussions(nextURL, headers, sendRequest));
+                }
+            }
+        }
+        return discussions;
+    }
+
+    private void deleteCommitDiscussionNote(String commitDiscussionNoteURL, Map<String, String> headers, boolean sendRequest) throws IOException {
+        //https://docs.gitlab.com/ee/api/discussions.html#delete-a-commit-thread-note
+        HttpDelete httpDelete = new HttpDelete(commitDiscussionNoteURL);
+        for (Map.Entry<String, String> entry : headers.entrySet()) {
+            httpDelete.addHeader(entry.getKey(), entry.getValue());
+        }
+
+        if (sendRequest) {
+            HttpResponse httpResponse = HttpClients.createDefault().execute(httpDelete);
+            validateGitlabResponse(httpResponse, 204, "Commit discussions note deleted");
+        }
+    }
+
+    private void postCommitComment(String commitCommentUrl, Map<String, String> headers, List<NameValuePair> params, boolean sendRequest) throws IOException {
+        //https://docs.gitlab.com/ee/api/commits.html#post-comment-to-commit
+        HttpPost httpPost = new HttpPost(commitCommentUrl);
+        for (Map.Entry<String, String> entry : headers.entrySet()) {
+            httpPost.addHeader(entry.getKey(), entry.getValue());
+        }
+        httpPost.setEntity(new UrlEncodedFormEntity(params));
+
+        if (sendRequest) {
+            LOGGER.info("Posting {} with headers {} to {}", params, headers, commitCommentUrl);
+
+            HttpResponse httpResponse = HttpClients.createDefault().execute(httpPost);
+            validateGitlabResponse(httpResponse, 201, "Comment posted");
+        }
+    }
+
+    protected void postStatus(String statusPostUrl, Map<String, String> headers, AnalysisDetails analysis, String comment, String coverage, boolean sendRequest) throws IOException{
+        //See https://docs.gitlab.com/ee/api/commits.html#post-the-build-status-to-a-commit
+        statusPostUrl += "?name=SonarQube";
+        String status = (analysis.getQualityGateStatus() == QualityGate.Status.OK ? "success" : "failed");
+        statusPostUrl += "&state=" + status;
+        statusPostUrl += "&target_url=" + URLEncoder.encode(String.format("%s/dashboard?id=%s&pullRequest=%s", server.getPublicRootUrl(),
+                URLEncoder.encode(analysis.getAnalysisProjectKey(),
+                        StandardCharsets.UTF_8.name()), URLEncoder
+                        .encode(analysis.getBranchName(),
+                                StandardCharsets.UTF_8.name())), StandardCharsets.UTF_8.name());
+        //statusPostUrl+="&description="+URLEncoder.encode(comment, StandardCharsets.UTF_8.name());
+        statusPostUrl+="&coverage="+coverage;
+        //TODO: add pipelineId if we have it
```

Comment:
Is this TODO still needed, and if so, when does this need done?

Reply:
Need to find out how I can get the environment variable `CI_PIPELINE_ID` from the scanner side. When the build is run with gilab ci and there are multiple pipelines for a MR, the result will then be attached to the correct pipeline.

Reply:
@mc1arke Do you have an idea how to get the value of an environment variable from the scanner back to the sonar server? Gitlab CI sets the `CI_PIPELINE_ID` variable and I would like to pass that down to the status posting. Otherwise the wrong pipeline could be annotated with the sonar result.

I couldn't find a way, so if you also don#t know anything, I will just remove the TODO comment.

Reply:
You won't be able to access the variable directly at this point as this is the Compute Engine executing on the Sonarqube server, rather than the scanner on the Gitlab pipeline. I don't think there would be an easy way to pass this information to the compute engine, but I'll see if there's any way to fudge it.